UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,736 @@
+
+        if (AsciiStrnCmp (NodePtr->Name, "upl-images@", AsciiStrLen ("upl-images@")) == 0) {
+          DEBUG ((DEBUG_INFO, "  Found image@ node \n"));
+          //
+          // Build PayloadBase HOB .
+          //
+          PayloadBase = BuildGuidHob (&gUniversalPayloadBaseGuid, sizeof (UNIVERSAL_PAYLOAD_BASE));
+          ASSERT (PayloadBase != NULL);
+          if (PayloadBase == NULL) {
+            return EFI_OUT_OF_RESOURCES;
+          }
+
+          PayloadBase->Header.Revision = UNIVERSAL_PAYLOAD_BASE_REVISION;
+          PayloadBase->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_BASE);
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "addr", &TempLen);
+
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data64       = (UINT64 *)(PropertyPtr->Data);
+            StartAddress = Fdt64ToCpu (*Data64);
+            DEBUG ((DEBUG_INFO, "\n         Property(00000000)  entry"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", StartAddress));
+
+            PayloadBase->Entry = (EFI_PHYSICAL_ADDRESS)StartAddress;
+          }
+        }
+
+        if (AsciiStrnCmp (NodePtr->Name, "upl-params", AsciiStrLen ("upl-params")) == 0) {
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "addr-width", &TempLen);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            DEBUG ((DEBUG_INFO, "\n         Property(00000000)  address_width"));
+            DEBUG ((DEBUG_INFO, "  %X", Fdt32ToCpu (*Data32)));
+            SizeOfMemorySpace = (UINT8)Fdt32ToCpu (*Data32);
+            BuildCpuHob (SizeOfMemorySpace, PcdGet8 (SizeOfIoSpace));
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "pci-enum-done", &TempLen);
+          if (TempLen > 0) {
+            Data16      = (UINT16 *)(PropertyPtr->Data);
+            PciEnumDone = (UINT8)Fdt16ToCpu (*Data16);
+            DEBUG ((DEBUG_INFO, "  Found PciEnumDone (%08X)\n", PciEnumDone));
+          }
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "boot-mode", &TempLen);
+          if (TempLen > 0) {
+            TempStr = (CHAR8 *)(PropertyPtr->Data);
+            if (AsciiStrCmp (TempStr, "normal") == 0) {
+              BootMode = BOOT_WITH_FULL_CONFIGURATION;
+            } else if (AsciiStrCmp (TempStr, "fast") == 0) {
+              BootMode = BOOT_WITH_MINIMAL_CONFIGURATION;
+            } else if (AsciiStrCmp (TempStr, "full") == 0) {
+              BootMode = BOOT_WITH_FULL_CONFIGURATION_PLUS_DIAGNOSTICS;
+            } else if (AsciiStrCmp (TempStr, "default") == 0) {
+              BootMode = BOOT_WITH_DEFAULT_SETTINGS;
+            } else if (AsciiStrCmp (TempStr, "s4") == 0) {
+              BootMode = BOOT_ON_S4_RESUME;
+            } else if (AsciiStrCmp (TempStr, "s3") == 0) {
+              BootMode = BOOT_ON_S3_RESUME;
+            }
+          }
+        }
+      }
+    }
+    // Optional
+    else if (AsciiStrnCmp (NodePtr->Name, "pci-rb", AsciiStrLen ("pci-rb")) == 0) {

```

Comment:
Can the above block that counts the number of root bridges be inlined here?

Reply:
We need to allocate HobDataSize basing on root bridge count and later parsing/filling root bridge. We can think about optimization later after this patch.

Reply:
Ack.

Reply:
defer to future patch.